Rook: Design document for admission controller support in rook (#5151)
rook/rook: design/ceph/admission-controller.md

> # Admission Controllers on Rook

Author: Fixed

# Background

This proposal is to add support for admission controllers in Rook. 

> Currently, user can manipulate CRD specs with any values which may result in application not functioning as expected. An admission controller is a piece of code that intercepts requests to the Kubernetes API server prior to persistence of the object, but after the request is authenticated and authorized. In that list, there are two special controllers: MutatingAdmissionWebhook and ValidatingAdmissionWebhook. These execute the mutating and validating (respectively) admission control webhooks which are configured in the API.

Author: Added valid reason to use admission controllers instead of OpenAPI validation

> The admission control process proceeds in two phases. In the first phase, mutating admission controllers are run. In the second phase, validating admission controllers are run. Note again that some of the controllers are both.

Author: Fixed

> If any of the controllers in either phase reject the request, the entire request is rejected immediately and an error is returned to the end-user.

Author: Added implementation and error handling parts. Let me know if anything needs to be added.

# Requirements

1. Webhook server which will validate the requests
2. TLS certificates for the server,
3. ValidatingWebhookConfig/MutatingWebhookConfig which will intercept requests and send a HTTPS request to the webhook server.
4. RBAC Components 

> # Admission Controllers in Rook

Author: Fixed

> # Admission Controllers in Rook

Author: I've added it as Admission controller concepts then added implementation details and setup in rook

> ![AC Image](design-pic.png)

Reviewer: ```suggestion
![AC Image](./design-pic.png)
```

Author: I have made changes to this. Images are not appearing. Probably some issue with the image markup or the image size ?

## Setup

### Certificate Management

> #### Development and Testing 

Author: Added commands to execute and verification Let me know if anything needs to added

> #### Development and Testing 

Author: Added.

![Dev_AC_Image](./ac_dev_design.png)

> We will be opting for using for self-signed certificates approved by the kubernetes cluster itself. Once common.yaml and the operator has been deployed.  Using scripts we will be performing the following tasks

Author: Removed filenames. Made generic. Added sample steps that script will perform

> * Creating the private key and certs using openssl with the names tls.key and tls.crt. We will be hardcoding these names for now as it will be explained in later section

Author: Fixed. Made generic

* Creating and sending a CSR to kubernetes for approval
* Once approved, a generic secret will be created with the given tls.key and tls.crt which will be later mounted onto the server pod for use.

> * Modifying the webhook.yaml to inject CA bundle onto the ValidatingWebhookConfig

Author: Fixed. Refered as generic configuration files

##### Openshift Container Platform

![Prod_AC_Image](rook ac-openshift.png)

In the OCP cluster, The management of certificates would be completely done by the Openshift Service CA Operator. 

On annotating any service with 
`service.beta.openshift.io/serving-cert-secret-name:<secret-name>` 

The operator automatically generates a signed serving certificate and key pair(tls.crt and tls.key) into a secret in the same namespace as the service. 
This secret will be mounted onto the webhook pod. Our server would be looking for the files (tls.key and tls.crt) and automatically set them for TLS communication which is why we had hardcoded earlier for dev and testing env.
For injecting a CA bundle onto the ValidatingWebhookConfig, we can annotate it with
`service.beta.openshift.io/inject-cabundle=true`

The operator on detecting this will set the ca bundle field in our config with a base64url-encoded CA signing bundle. These certificates will be automatically rotated and renewed by the service CA operator. This can also be done manually through the commands which can be found in the operator manual.

> **All the annotated components will be provided in a separate file called openshift-webhook.yaml**

Author: Added reference links plus bunch of changes to make it generic. Let me know if any other changes are required

# Admission Controllers in Rook

# Background

This proposal is to add support for admission controllers in Rook. An admission controller is a piece of code that intercepts requests to the Kubernetes API server prior to persistence of the object, but after the request is authenticated and authorized

 There are two special controllers: MutatingAdmissionWebhook and ValidatingAdmissionWebhook.
 
 Mutating controllers may modify the objects they admit but validation controllers are only allowed to validate requests.

> Currently, user can manipulate CRD specs with any values which may result in Rook not functioning as expected. The present validation method in Kubernetes is the OpenAPIValidation schema validation which can be used for basic validations like checking type of data, providing a range for the values etc but anything more complex (checking resource availability, network status, error handling) would not be possible under this scenario.

Reviewer 2: ```suggestion
Currently, user can manipulate Custom Resource specs with any values which may result in Rook not functioning as expected. The present validation method in Kubernetes is the OpenAPIValidation schema validation which can be used for basic validations like checking type of data, providing a range for the values etc but anything more complex (checking resource availability, network status, error handling) would not be possible under this scenario.
```

Author: Fixed

> Currently, user can manipulate CRD specs with any values which may result in Rook not functioning as expected. The present validation method in Kubernetes is the OpenAPIValidation schema validation which can be used for basic validations like checking type of data, providing a range for the values etc but anything more complex (checking resource availability, network status, error handling) would not be possible under this scenario.

Reviewer 2: ```suggestion
Currently, user can manipulate CRD specs with any values which may result in Rook not functioning as expected. The present validation method in Kubernetes is the OpenAPI schema validation which can be used for basic validations like checking type of data, providing a range for the values etc but anything more complex (checking resource availability, network status, error handling) would not be possible under this scenario.
```

Author: Fixed

 
# Admission Controllers Concept

![AC Image](ac_design.png)

## Requirements

1. Webhook server which will validate the requests
2. TLS certificates for the server,
3. ValidatingWebhookConfig/MutatingWebhookConfig which will intercept requests and send a HTTPS request to the webhook server.
4. RBAC Components 

## Implementation 

As shown in the above diagram, the admission control process proceeds in two phases. In the first phase, mutating admission controllers are run. In the second phase, validating admission controllers are run. Note again that some of the controllers are both.

> The admission controllers intercept requests based on the values given in the configuration.In this config, We have to provide the details on 

Reviewer 2: ```suggestion
The admission controllers intercept requests based on the values given in the configuration. In this config, we have to provide the details on 
```

Author: Fixed

* What resources should it be looking for ? (Pods, Service)
* What api version and group does it belong to ? *Example : ApiVersion = (v1, v1beta) Group version = (rook.ceph.io, admissionregistration.k8s.io)*
* What kind of operation should it intercept ? (Create, Update, Delete)
* A valid base64 encoded CA bundle.
* What path do we want to send with HTTPs request (/validate, /mutate)

> A server should be in place (*with valid TLS certificates*) to intercept any HTTPs request that comes with the above path value. Once intercepted by a server, an [AdmissionRequest](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L40) object is sent through with the resource specifications. 

Reviewer 2: ```suggestion
A webhook server should be in place (*with valid TLS certificates*) to intercept any HTTPs request that comes with the above path value. Once the request is intercepted by the server, an [AdmissionRequest](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L40) is sent through with the resource specifications. 
```

Author: Fixed

> Here is the part where the server will perform all our validations on the provided resources values and send back an [AdmissionResponse](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L116) with the field Allowed set accordingly to indicate whether it is accepted or rejected.

Reviewer 2: This part seems a little misleading. Suggesting changes,
```suggestion
When the webhook server receives Admission Request, it will perform predefined validations on the provided resource values and send back an [AdmissionResponse](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L116) with indication to whether request is accepted or rejected.
```

Author: Fixed

If any of the controllers in either phase reject the request, the entire request is rejected immediately and an error is returned to the end-user.

## Certificate Management

### Development and Testing 

> ![Dev_AC_Image](./ac_dev_design.png)

Reviewer 2: This diagram is incomplete,
1. MutatingWebhookConfig is not shown.
2. `/validate` endpoint is called twice. It should be once per request.
3. Terms mentioned in boxes like `service` and `webhook` should be self-explanatory.

Author: 1. Added MutatingWebhookConfig label on the same as ValidatingWebhookConfig as both of them are doing the same, just sending different paths
2. /validate request would be sent once to our kubernetes service which in turn would pass it to our server. No request would be sent twice. Thanks for this suggestion, will make it more clear.
3. I have labeled service as Kubernetes service endpoint and webhook as HTTPS Webhook Server to make it more clear. Let me know if you have any other suggestions

* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
* All the above resources will be created in rook-ceph namespace

Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.

Below is an excerpt of what a ValidatingWebhookConfig looks like 

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:
  name: demo-webhook
webhooks:
  - name: webhook-server.webhook-demo.svc
    clientConfig:
      service:
        name: webhook-server
        namespace: rook-ceph
        path: "/validate"
      caBundle: ${CA_PEM_B64}
    rules:
      - operations: [ "CREATE" ]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
```

We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.

## Setup in Rook

### Starting the Webhook Server

 Based on the whether the  secrets are present, the rook operator will  deploy the relevant configuration files onto the cluster and start the server. 

> The secrets will be volume mounted on the rook operator pod dynamically when they are detected. After the volumes are mounted, an HTTP server would be started.

Reviewer 2: ```suggestion
The secrets will be volume mounted on the rook operator pod dynamically when they are detected. After the volumes are mounted, an HTTPs server would be started.
```

Author: Fixed

Once the server starts, it will look for the appropriate tls key and crt files in the mounted volumes and start intercepting requests based on the path set in ValidatingWebhookConfig.

### Error Handling

> If the server is unable to find the proper secrets certificates, It will simply not deploy any admission controller components onto the cluster and hence rook will continue to function normally as before.

Reviewer 2: ```suggestion
If the server is unable to find valid certificates, it will not deploy any admission controller components onto the cluster and hence rook will continue to function normally as before.
```

Author: Fixed

# Admission Controllers in Rook

# Background

This proposal is to add support for admission controllers in Rook. An admission controller is a piece of code that intercepts requests to the Kubernetes API server prior to persistence of the object, but after the request is authenticated and authorized

 There are two special controllers: MutatingAdmissionWebhook and ValidatingAdmissionWebhook.
 
 Mutating controllers may modify the objects they admit but validation controllers are only allowed to validate requests.

Currently, user can manipulate CRD specs with any values which may result in Rook not functioning as expected. The present validation method in Kubernetes is the OpenAPIValidation schema validation which can be used for basic validations like checking type of data, providing a range for the values etc but anything more complex (checking resource availability, network status, error handling) would not be possible under this scenario.
 
# Admission Controllers Concept

> ![AC Image](./design-pic.png)

Reviewer 3: nit: how about adding the `ac-` prefix to the png filename?

Author: Fixed

## Requirements

1. Webhook server which will validate the requests
2. TLS certificates for the server,
3. ValidatingWebhookConfig/MutatingWebhookConfig which will intercept requests and send a HTTPS request to the webhook server.
4. RBAC Components 

## Implementation 

As shown in the above diagram, the admission control process proceeds in two phases. In the first phase, mutating admission controllers are run. In the second phase, validating admission controllers are run. Note again that some of the controllers are both.

The admission controllers intercept requests based on the values given in the configuration.In this config, We have to provide the details on 
* What resources should it be looking for ? (Pods, Service)
* What api version and group does it belong to ? *Example : ApiVersion = (v1, v1beta) Group version = (rook.ceph.io, admissionregistration.k8s.io)*
* What kind of operation should it intercept ? (Create, Update, Delete)
* A valid base64 encoded CA bundle.
* What path do we want to send with HTTPs request (/validate, /mutate)

> A server should be in place (*with valid TLS certificates*) to intercept any HTTPs request that comes with the above path value. Once intercepted by a server, an AdmissionRequest(https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L40) object is sent through with the resource specifications. 

Reviewer 3: ```suggestion
A server should be in place (*with valid TLS certificates*) to intercept any HTTPs request that comes with the above path value. Once intercepted by a server, an [AdmissionRequest](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L40) object is sent through with the resource specifications. 
```

Author: Fixed

Here is the part where the server will perform all our validations on the provided resources values and send back an AdmissionResponse with the field Allowed set accordingly to indicate whether it is accepted or rejected.

Here is structure of our response object. 

**AdmissionResponse Object**
```

> type AdmissionResponse struct {

Reviewer 3: We're not defining this type, right? Can we link to a doc or source file where it is defined instead? Then it doesn't look like we're defining the response type.

Author: Added links for the struct

	// AuditAnnotations is an unstructured key value map set by remote admission controller (e.g. error=image-blacklisted).
	// MutatingAdmissionWebhook and ValidatingAdmissionWebhook admission controller will prefix the keys with
	// admission webhook name (e.g. imagepolicy.example.com/error=image-blacklisted). AuditAnnotations will be provided by
	// the admission webhook to add additional context to the audit log for this request.
	// +optional
	AuditAnnotations map[string]string `json:"auditAnnotations,omitempty" protobuf:"bytes,6,opt,name=auditAnnotations"`
}
```

If any of the controllers in either phase reject the request, the entire request is rejected immediately and an error is returned to the end-user.

## Certificate Management

### Development and Testing 

![Dev_AC_Image](./ac_dev_design.png)

We can use self-signed certificates approved by the Kubernetes Certificate Authority for development purposes.This can be done by following the steps given below.
* Creating the private key and certs using openssl.

**Sample : Generating the public and private keys** 

```
openssl genrsa -out ${PRIVATE_KEY_NAME}.pem 2048

>  openssl req -new -key ${PRIVATE_KEY_NAME}.pem -subj "/CN=${service}.${namespace}.svc"

Reviewer 3: ```suggestion
openssl req -new -key ${PRIVATE_KEY_NAME}.pem -subj "/CN=${service}.${namespace}.svc"
```

Author: Fixed

kubectl create secret generic ${secret} \
        --from-file=key.pem=${PRIVATE_KEY_NAME}.pem \
        --from-file=cert.pem=${PUBLIC_KEY_NAME}.pem
```
* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
* All the above resources will be created in rook-ceph namespace

Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.

Below is an excerpt of what a ValidatingWebhookConfig looks like 

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:
  name: demo-webhook
webhooks:
  - name: webhook-server.webhook-demo.svc
    clientConfig:
      service:
        name: webhook-server
        namespace: rook-ceph
        path: "/validate"
      caBundle: ${CA_PEM_B64}
    rules:
      - operations: [ "CREATE" ]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
```

We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.

## Setup in Rook

### Starting the Webhook Server

> The webhook server binaries will be bundled with rook operator image. Based on the whether the relevant secrets are present, the rook operator will automatically trigger the webhook binaries, deploy the relevant configuration files onto the cluster and start the server. 

Reviewer 3: When you have a prototype of this I'm curious how much the image size will increase with the webhook binaries. I would have hoped Rook would just be referencing libraries for starting the https server and there wouldn't be new binaries to pull into the image.

Author: There would definitely be an increase an image size we may want to avoid.  I have modified my statement here. The rook operator itself would be starting an HTTP server (through libraries) when the appropriate secrets are detected.  Thanks for your suggestion

	// +optional
	Result *metav1.Status `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`

	// The patch body. Currently we only support "JSONPatch" which implements RFC 6902.
	// +optional
	Patch []byte `json:"patch,omitempty" protobuf:"bytes,4,opt,name=patch"`

	// The type of Patch. Currently we only allow "JSONPatch".
	// +optional
	PatchType *PatchType `json:"patchType,omitempty" protobuf:"bytes,5,opt,name=patchType"`

	// AuditAnnotations is an unstructured key value map set by remote admission controller (e.g. error=image-blacklisted).
	// MutatingAdmissionWebhook and ValidatingAdmissionWebhook admission controller will prefix the keys with
	// admission webhook name (e.g. imagepolicy.example.com/error=image-blacklisted). AuditAnnotations will be provided by
	// the admission webhook to add additional context to the audit log for this request.
	// +optional
	AuditAnnotations map[string]string `json:"auditAnnotations,omitempty" protobuf:"bytes,6,opt,name=auditAnnotations"`
}
```

If any of the controllers in either phase reject the request, the entire request is rejected immediately and an error is returned to the end-user.

## Setup in Rook

> ### Certificate Management

Reviewer 3: Since cert management is not managed by Rook, how about not putting this under the "Setup in Rook" section?

Author: Fixed. Added "Starting the webhook server" part under "Setup in Rook"

#### Development and Testing 

![Dev_AC_Image](./ac_dev_design.png)

> We will be opting for using for self-signed certificates approved by the kubernetes cluster itself. Once common.yaml and the operator has been deployed.  Using scripts we will be performing the following tasks

Reviewer 3: Can't the cert management be done as a step before common.yaml is created? If you just need the namespace to exist, you could create that manually with the cert management script. Then we could keep it independent from common.yaml.

Author: We could create namespace using the scripts but then later on deploying common.yaml may cause namespace already existing error. I just added it this way to keep the complete deployment clean without any errors. Please let me know if you would prefer having the namespace created in the scripts

```
* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
* All the above resources will be created in rook-ceph namespace

Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.

Below is an excerpt of what a ValidatingWebhookConfig looks like 

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:
  name: demo-webhook
webhooks:
  - name: webhook-server.webhook-demo.svc
    clientConfig:
      service:
        name: webhook-server
        namespace: rook-ceph
        path: "/validate"
      caBundle: ${CA_PEM_B64}
    rules:
      - operations: [ "CREATE" ]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
```

We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.

#### Production

The onus of providing proper certificates and CA bundle for the server would be on the user. The following example would provide how it would work in an OCP cluster environment

> ##### Openshift Container Platform

Reviewer 2: This section looks like a candidate for Documentation and not design doc.

> ##### Openshift Container Platform

Reviewer 2: This section looks like a candidate for Documentation and not design doc.

Author: @umangachapagain What changes do you suggest for this ?

Author: I have segregated the PR into 2 commits. One commit is for the design documentation and the 2nd commit adds docuemntation for deploying in OCP

> ![Prod_AC_Image](./rook ac-openshift.png)

Reviewer 3: nit: for consistency how about naming the image `ac-openshift.png`

Author: Fixed

As shown in the above diagram, In the OCP cluster, The management of certificates would be completely done by the Openshift Service CA Operator. 

The first step would be to generate the certificates and assign them to the proper components. On annotating any service with 

`service.beta.openshift.io/serving-cert-secret-name:<secret-name>` 

We have to deploy the the annotated service at the start to generate the signed certificates. This can done manually  through the command 

`oc create -f ${SERVICE_NAME}.yaml`

For verifying 

`oc describe ${SERVICE_NAME} -n ${NAMESPACE}`

**Output**
```
oc describe service <service-name>
...
Annotations:              service.beta.openshift.io/serving-cert-secret-name: <service-name>
                          service.beta.openshift.io/serving-cert-signed-by: openshift-service-serving-signer@1556850837
```

> The operator automatically generates a signed serving certificate and key pair into a secret in the same namespace as the service. 

Reviewer 3: I thought the operator was just going to look for the cert, rather than generating it? Did you mean the script will do this? Or did you say the operator can only do this for OpenShift and not K8s?

Author: Sorry for being unclear, I mean to say that the Openshift Service CA operator is going to be the one which generate and maintain the certificates (https://docs.openshift.com/container-platform/4.1/authentication/certificates/service-serving-certificate.html#understanding-service-serving_service-serving-certificate) . This operator works as a part of OCP and would work only for Openshift not K8s. I will modify the statement here.

This secret will be mounted onto the server pod. Our server would be looking for tls certificates and automatically set them for TLS communication.
For injecting a CA bundle onto the ValidatingWebhookConfig, we can annotate it with

`service.beta.openshift.io/inject-cabundle=true`

The operator on detecting this will set the ca bundle field in our config with a base64url-encoded CA signing bundle. These certificates will be automatically rotated and renewed by the service CA operator. This can also be done manually through the commands which can be found in the operator manual.

> ### Starting the Webhook Server

Reviewer 3: This is where the "Rook" heading should be, right?

Author: Fixed. Added "Starting the webhook server" part under "Setup in Rook"

The webhook server binaries will be bundled with rook operator image. Based on the whether the relevant secrets are present, the rook operator will automatically trigger the webhook binaries, deploy the relevant configuration files onto the cluster and start the server. 

> After the start of the server, it will look for the appropriate tls key and crt files and start intercepting requests based on the path set in ValidatingWebhookConfig.

Reviewer 3: How about documenting where the operator will look for the key?

Author: Added content on this ....

The admission control process proceeds in two phases. In the first phase, mutating admission controllers are run. In the second phase, validating admission controllers are run. Note again that some of the controllers are both.

If any of the controllers in either phase reject the request, the entire request is rejected immediately and an error is returned to the end-user.

# Requirements

1. Webhook server which will validate the requests
2. TLS certificates for the server,
3. ValidatingWebhookConfig/MutatingWebhookConfig which will intercept requests and send a HTTPS request to the webhook server.
4. RBAC Components 

#Admission Controllers in Rook

![AC Image](./design-pic.png)

##Setup
###Certificate Management
####Development and Testing 
![AC Image](./ac_dev_design.png)

We will be opting for using for self-signed certificates approved by the kubernetes cluster itself. Once common.yaml and the operator has been deployed.  Using scripts we will be performing the following tasks

> * Creating the private key and certs using openssl with the names tls.key and tls.crt. We will be hardcoding these names for now as it will be explained in later section

Reviewer 3: The filenames are just for working with the script, right?

Author: Actually in the server code, we are looking for these files specifically to complete the configuration hence if these names are changed, the server would throw an error.

* Creating and sending a CSR to kubernetes for approval

> * Once approved, a generic secret will be created with the given tls.key and tls.crt which will be later mounted onto the server pod for use.

Reviewer 3: What is required for approval? Is this a manual step by the admin, or can it be automated by the script if the admin is running the script?

Author: Yes Travis, thats how it goes. The scripts will do all the heavy lifting of sending the CSR and running command for approval. Admins won't have to do anything.  Please check the PR link in have provided in the description, you could see the implementation.

Author: Added steps for commands and verification

* Modifying the webhook.yaml to inject CA bundle onto the ValidatingWebhookConfig

> * Creating a webhook deployment, service, config on the cluster and deploying the entire webhook setup

Reviewer 3: Won't this step be done by the operator?

Author: Yes, I would like to have a discussion on it today if possible. I have a few concerns regarding how it will work in dev/test and openshift prod.

Author: Fixed from recent design decisions

* Now our webhook server has been configured, deployed and ready to intercept requests based on the webhookconfig.

> * **All the above resources will be created in rook-ceph namespace hence it should be created in prior**

Reviewer 3: > it should be created in prior

What do you mean? The rook-ceph namespace just already needs to be created first?

Author: Yes travis, as all the components will be created in rook-ceph namespace. The namespace would need to have been already created but since the operator would be creating the admission controller, i guess the the namespace would already exist. As we are doing this from this scripts for dev/testing purpose, i thought it be best to add that point.  I'll be modifying it !

Author: Fixed

Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.

Below is an excerpt of what a ValidatingWebhookConfig looks like 

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:
  name: demo-webhook
webhooks:
  - name: webhook-server.webhook-demo.svc
    clientConfig:
      service:
        name: webhook-server
        namespace: rook-ceph
        path: "/validate"
      caBundle: ${CA_PEM_B64}
    rules:
      - operations: [ "CREATE" ]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
```

We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.

> ####Production

Reviewer 3: nit: Some of the titles like this one are not rendered correctly. They need a blank line before and after the title, and a space between the hashes and the words

Author: Yes, i noticed images are not appearing properly. I have applied your changes (They need a blank line before and after the title, and a space between the hashes and the words). Not sure what is going wrong there.

On annotating any service with 
`service.beta.openshift.io/serving-cert-secret-name:<secret-name>` 

The operator automatically generates a signed serving certificate and key pair(tls.crt and tls.key) into a secret in the same namespace as the service. 
This secret will be mounted onto the webhook pod. Our server would be looking for the files (tls.key and tls.crt) and automatically set them for TLS communication which is why we had hardcoded earlier for dev and testing env.
For injecting a CA bundle onto the ValidatingWebhookConfig, we can annotate it with
`service.beta.openshift.io/inject-cabundle=true`

The operator on detecting this will set the ca bundle field in our config with a base64url-encoded CA signing bundle. These certificates will be automatically rotated and renewed by the service CA operator. This can also be done manually through the commands which can be found in the operator manual.

**All the annotated components will be provided in a separate file called openshift-webhook.yaml**

###Starting the Webhook Server

The webhook server binaries will be bundled with rook operator image. An environment variable will be set in rook operator

> ```ROOK_ADMISSION_CONTROLLER : TRUE```

Reviewer 3: Instead of adding a new setting like this, what if the operator automatically started the admission controller if it finds the expected secret with the tls key?

Author: That is good idea. I would like to have a few more discussions on this today. I'll schedule a call.

Author: Made the changes for this in the design doc. Please review
